ZEP000-build-state-on-followers.md

```diff
@@ -0,0 +1,508 @@
+
+They all have access to the records (documents), and the Leader worker is building a state based on the already processed records. He will regularly share the state of his already processed documents with the others.
+
+![stateDocuments](images/stateDocuments.png)
+
+If now the Leader is on vacation or sick someone else needs to take over. This might imply several problems.
+
+![backupWorkers](images/backupWorkers.png)
+
+If the Leader has forgotten to send the state before he left, one of the office worker, which takes over, might have a quite old state, and he needs to do a lot of rework. If the state is not readable he has also a problem to take over or there are documents missing etc.
+
+In order to prevent such situations the idea is that each office worker, since they already have access to the documents (records), are building their own state. The Leader is only in charge of sending letters, invoices out etc. If now the Leader is not available another worker can take immediately over. To prevent not answering letters or answering twice, after take over, the Leader will mark documents as processed. The other workers will only build their state on already processed documents.
+
+## Distinction to normal raft
+
+As you can read in the [raft paper](https://raft.github.io/raft.pdf) chapter 7, compaction and taking a snapshot should be done on each server.
+
+> This snapshotting approach departs from Raft’s strong
+leader principle, since followers can take snapshots without the knowledge of the leader. However, we think this
+departure is justified. While having a leader helps avoid
+conflicting decisions in reaching consensus, consensus
+has already been reached when snapshotting, so no decisions conflict. Data still only flows from leaders to fol
+lowers, just followers can now reorganize their data.
+>
+> We considered an alternative leader-based approach in
+which only the leader would create a snapshot, then it
+would send this snapshot to each of its followers. However, this has two disadvantages. First, sending the snapshot to each follower would waste network bandwidth and
+slow the snapshotting process. Each follower already has
+the information needed to produce its own snapshots, and
+it is typically much cheaper for a server to produce a snapshot from its local state than it is to send and receive one
+over the network. Second, the leader’s implementation
+would be more complex. For example, the leader would
+need to send snapshots to followers in parallel with replicating new log entries to them, so as not to block new
+client requests.
+
+As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.
+
+Unfortunately our processing state machine produces new commands, which need to be applied to the state machine again. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records.
```

Comment:
In normal RAFT system you have one statemachine which you can run on all machines, no matter what type/role the node has. In our case this is not true, since only the leader should create follow up events. So we have a distinction between processing and replay state machine. The first one runs on the leader the other on the follower.

Reply:
But how does that stop us from **making snapshots on followers**?

Reply:
Why does it stop us? The text just says that we can;t just run the same state machine, the same logic to build it.

Reply:
Ah I see, the problem is that you're describing 2 different topics in this subchapter.

The heading of this subchapter is `Distinction to normal raft`. If we can make snapshots on followers, then that's actually the same as in the paper right? Perhaps it just doesn't belong here, but in a separate heading called `Snapshots on followers`.

Reply:
> If we can make snapshots on followers, then that's actually the same as in the paper right? 

It is not the same, since we can't just run the same code which we run on the Leader :)

Reply:
I can also remove it  when it lead to more confusion than it helps. Any opinion @deepthidevaki

Reply:
Well it makes sense to discuss the topic of snapshots on followers to some extend. Is it in scope of this ZEP or out of scope? If it's out of scope you could also mention it in that section,

Reply:
I tried to rewrite it to make it more clear. Does that help?

```
As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.

However, we cannot run the same state machine on the leader and followers as in the original raft.
Our processing state machine produces new follow up commands.The order in which follow up commands and events are written to the logstream in relation to concurrent external commands is decided by the leader. When the follower runs the same state machine, it will also generate new follow up commands and events, which might result in a different order of records. This means that we can't just run the same processing state machine on all nodes, since only the leader should be able to create new records. As later explained in this document, the leader's state machine process commands and generate followup commands and events. Follower runs a replay state machine where it only applies the events generated by the leader.
```

Reply:
I still don't see what this has to do with snapshots on followers. It seems to describe the difference between a leader (processes: writes follow-up records) and a follower (only replays), but it does not discuss the impact on snapshots on those followers. 

At this point it is still unclear to me whether snapshots on followers are in scope of the ZEP or not

Reply:
I have the feeling you are missing some understanding of the ZEP. In general the ZEP is about building state on Followers, in order to do that we replay events on the followers. Taking snapshot on Followers is just a part of it.

Reply:
I'm wondering why this is the case. What I have missed to make this clear? Or what do you think we should add?

Reply:
I understand that the ZEP is about state on followers. and so it also makes sense to discuss snapshots because of the reduced complexity and network traffic related to distributing snapshots if you only make them on leaders.

To me the part up to and including the sentence `As the paper states it makes totally sense to build state and take snapshots of it on all servers unrelated of the raft state.` was clear. 

The paragraph under it, then continues with `Unfortunately` implying that even though it seems possible to create snapshots on followers, it is not. And then continues to talk about the difference between leaders and followers concerning the difference of processing and replay without actually stating how this affects snapshots. This paragraph just confuses me about what was mentioned before it. And I think the reason is that they just discuss different topics (as far as I understand now, the last paragraph has nothing to do with snapshots)

Reply:
Sorry to be so difficult 😄 I just want to understand it

Reply:
To take snapshot on followers, we need to build state on followers. To build state on followers, we need to run a state machine on the followers. If we could follow the normal raft approach, we can run the same state machine on the leader and follower. Then we don't need this ZEP, as it is trivial to build state and take snapshots. So the main difference compared to raft is how to build state on followers. Once we know how to build state on followers, taking snapshot is not difficult.

Reply:
Thanks @deepthidevaki that is exactly it

Reply:
Thanks @deepthidevaki that makes sense. Can we add this to this subchapter?

Reply:
If you think it helps sure. I will change and apply all the hints after all the reviews are done, such that the other reviews haven't any conflicts on reading.

Reply:
https://github.com/zeebe-io/enhancements/pull/16/commits/c521577fa1cea096df23157ea95af55c5d43b10d